File: Solutions/LowestCommonAncestorOfABinarySearchTree.cpp
// Leetcode #235: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// Solution 1 | O(logn) time, O(1) space
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        // Store the smaller and larger value of nodes p and q
        int minVal = min(p->val, q->val), maxVal = max(p->val, q->val);
        
        // LCA must be between [minVal, maxVal] inclusive
        // While the node's is NOT between p and q, we need to go lower
        TreeNode* curr = root;
        while (!(curr->val >= minVal && curr->val <= maxVal)) {
            // If the node is too big, go left. Else, go right.
            curr = (curr->val > maxVal) ? curr->left : curr->right;
        }
            
        return curr; // Return the LCA
    }
};

File: Solutions/LowestCommonAncestorOfABinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# Solution 1 | O(logn) time, O(1) space
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # Get the smaller and larger values of p and q
        smallerVal, biggerVal = sorted([p.val, q.val])

        # Starting from the root node, search for LCA(p, q)
        # LCA(p, q) must have a value between p and q
        while (not smallerVal <= root.val <= biggerVal):
            # If root.val is too big, both p and q are on the right side. Vice versa
            root = root.left if root.val > smallerVal else root.right

        return root
